cli/state.rs

```diff
@@ -174,13 +152,14 @@ impl Loader for ThreadSafeState {
   ) -> Box<deno::SourceCodeInfoFuture> {
     self.metrics.resolve_count.fetch_add(1, Ordering::SeqCst);
     Box::new(
-      fetch_module_meta_data_and_maybe_compile_async(self, module_specifier)
-        .map(|module_meta_data| deno::SourceCodeInfo {
-          // Real module name, might be different from initial URL
+      fetch_source_file_and_maybe_compile_async(self, module_specifier).map(
+        |source_file| deno::SourceCodeInfo {
+          // Real module name, might be different from initial specifier
           // due to redirections.
-          code: module_meta_data.js_source(),
-          module_name: module_meta_data.module_name,
-        }),
+          code: source_file.js_source(),
+          module_name: source_file.url.to_string(),
```

Comment:
"Real module name, might be different from initial specifier" --> I know this is not something you introduced, and I'm not asking you to change it now.
But I think that, instead of making this two-URLs idea pervasive throughout the code, we could just deal with this at the level of the Loader/Modules object.
Maybe it'd be better to make `fetch_source_file()` return `enum { Code(SourceCodeInfo), Redirect(Url) }` or something like that.
What do you think @bartlomieju ?
(again - i'm not asking to make that part of this patch, just looking for input)

Reply:
Glad you're raising this. 

AFAIK `core/modules.rs` already handles situation with different URL than requested and registers module under aliases in that situation ([ref](https://github.com/denoland/deno/blob/1fde15c0bc77e0c5cc6f7cbf27b4e072e1ad6c46/core/modules.rs#L189-L215)).

On the other hand I believe we're gonna abstract away compilers even more (eg. `JsonCompiler` might be a good idea to abstract away wrapping JSON files in default export) and it should be theirs responsibility to handle all redirections for source files.

Also, please take a look at this comment: https://github.com/denoland/deno/pull/2636#discussion_r303029242

Reply:
After second thought - I think current solution where module is requested under one URL and might get different is better. It completely decouples redirections from V8 and leave it to `DenoDir` to handle that.

Reply:
I don't disagree, but just I just want to note that there is kind of some code for redirect already in core
https://github.com/denoland/deno/blob/1fde15c0bc77e0c5cc6f7cbf27b4e072e1ad6c46/core/modules.rs#L280

Reply:
> After second thought - I think current solution where module is requested under one URL and might get different is better. It completely decouples redirections from V8 and leave it to `DenoDir` to handle that.

I don't think that works, because if the redirected module imports other modules, they are resolved with the *redirected* url as a base. At least that's how I think HTTP semantics work and what browsers would do. So at a minimum the Modules map needs to know about both the internal and external URL.

This is why I am suggesting to make Modules fully responsible, and make it so that the fetch/cache/compile infrastructure doesn't handle redirects, it only forwards the new url to Modules.

Reply:
> I don't think that works, because if the redirected module imports other modules, they are resolved with the _redirected_ url as a base. At least that's how I think HTTP semantics work and what browsers would do. So at a minimum the Modules map needs to know about both the internal and external URL.

That's a good scenario. I think I once tested for this, but can't remember exactly how it behaved. Let me get back to you on this one

Reply:
@piscisaureus I did some tests with Deno and Chrome using `//tests/026_redirect_javascript.js` file and it looks like behaviours match.

After first redirect from `:4567` to `:4545` subsequent module is requested using referrer address after redirection.


**Output from Deno**
```
... snip ...
DEBUG RS - deno_bin:339 - main_module file:///Users/biwanczuk/dev/deno/tests/026_redirect_javascript.js
DEBUG RS - deno_bin::deno_dir:279 - fetch_source_file. specifier file:///Users/biwanczuk/dev/deno/tests/026_redirect_javascript.js 
DEBUG RS - deno::modules:393 - register_complete file:///Users/biwanczuk/dev/deno/tests/026_redirect_javascript.js
DEBUG RS - deno_bin::deno_dir:279 - fetch_source_file. specifier http://localhost:4547/redirects/redirect3.js 
DEBUG RS - deno_bin::deno_dir:253 - deps filename: "/Users/biwanczuk/Library/Caches/deno/deps/http/localhost_PORT4547/redirects/redirect3.js"
... snip ...
DEBUG RS - deno_bin::http_util:102 - Redirecting to http://localhost:4545/tests/subdir/redirects/redirect3.js...
DEBUG RS - deno_bin::deno_dir:253 - deps filename: "/Users/biwanczuk/Library/Caches/deno/deps/http/localhost_PORT4545/tests/subdir/redirects/redirect3.js"
... snip ...
DEBUG RS - deno::modules:393 - register_complete http://localhost:4545/tests/subdir/redirects/redirect3.js
DEBUG RS - deno_bin::deno_dir:279 - fetch_source_file. specifier http://localhost:4545/tests/subdir/redirects/redirect1.js 
DEBUG RS - deno_bin::deno_dir:253 - deps filename: "/Users/biwanczuk/Library/Caches/deno/deps/http/localhost_PORT4545/tests/subdir/redirects/redirect1.js"
... snip ...
DEBUG RS - deno::modules:393 - register_complete http://localhost:4545/tests/subdir/redirects/redirect1.js
3 imports 1
```

**Output from Chrome network panel** - trimmed down HAR file
```
... snip ...
[
      {
        "request": {
          "method": "GET",
          "url": "http://0.0.0.0:4500/tests/026_redirect_javascript.js",
          ... snip ...
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          ... snip ...
        },
      },
      {
        "request": {
          "method": "GET",
          "url": "http://localhost:4547/redirects/redirect3.js",
        },
        "response": {
          "status": 301,
          "statusText": "Moved Permanently",
          "httpVersion": "http/1.0",
          "headers": [
            {
              "name": "Location",
              "value": "http://localhost:4545/tests/subdir/redirects/redirect3.js"
            },
          ],
          ... snip ...
        },
      },
      {
        "request": {
          "method": "GET",
          "url": "http://localhost:4545/tests/subdir/redirects/redirect3.js",
          "httpVersion": "http/1.0",
          ... snip ...
          ],
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          ... snip ...
        },
      },
      {
        "request": {
          "method": "GET",
          "url": "http://localhost:4545/tests/subdir/redirects/redirect1.js",
          "httpVersion": "http/1.0",
          "headers": [
            {
              "name": "Referer",
              "value": "http://localhost:4545/tests/subdir/redirects/redirect3.js"
            },
            ... snip ...
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          "httpVersion": "http/1.0",

      }
    ]
  }
}
```

Reply:
@bartlomieju Cool, thanks for digging into that, is good to see this confirmed.